### PR TITLE
Fix cloudpickle incompatibilities on early Python 3.5 versions

### LINKED ISSUE #360: 1.4.0 doesn't support python 3.5.2 and below
Since cloudpickle 1.4.0 uses ClassVar from [this commit](https://github.com/cloudpipe/cloudpickle/commit/1ba10a5491d3f7b788d36c0d3cacc77bd41668a0#diff-d2a3618afedd4e124c532151eedbae09R64), which is only supported by [Python 3.5.3 and above](https://docs.python.org/3/library/typing.html#typing.ClassVar).

I'm amateur python developer, so I'm not sure of the best way to address / announce this. But this may be worth fixing because Ubuntu 16.04 ships with python 3.5.1-2, so all users of that version of Ubuntu would be affected as well.

Thanks!

### PR DESCRIPTION
**Pull request by pierreglaser**
Closes #360 . `cloudpickle 1.4.0` is not compatible with early `Python 3.5` versions.
This should fix it.

Note that I did not set up any CI for `Python 3.5.[0-2]`, I simply tested it on my machine using fresh `conda` envs.

@vedran If you have some time, could you tell me if this branch fixes the problems that made you create #360?

I would be tempted to release a bugfix version by tonight since this bug completely breaks `cloudpickle` on `Python 3.5`.

### CODE REVIEW

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -427,10 +432,24 @@ def _extract_class_dict(cls):
 if sys.version_info[:2] < (3, 7):  # pragma: no branch
     def _is_parametrized_type_hint(obj):
         # This is very cheap but might generate false positives.
-        origin = getattr(obj, '__origin__', None)  # typing Constructs
-        values = getattr(obj, '__values__', None)  # typing_extensions.Literal
-        type_ = getattr(obj, '__type__', None)     # typing_extensions.Final
-        return origin is not None or values is not None or type_ is not None
+        # genral typing Constructs
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:435]:
```suggestion
        # general typing Constructs
```

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -966,14 +985,32 @@ def _save_parametrized_type_hint(self, obj):
                 initargs = (Final, obj.__type__)
             elif type(obj) is type(ClassVar):
                 initargs = (ClassVar, obj.__type__)
-            elif type(obj) in [type(Union), type(Tuple), type(Generic)]:
-                initargs = (obj.__origin__, obj.__args__)
+            elif type(obj) is type(Generic):
+                if sys.version_info < (3, 5, 2):  # pragma: no cover
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:989]: Maybe it's possible to ask for a specific minor version on github actions?

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -427,10 +432,24 @@ def _extract_class_dict(cls):
...
+
+        # typing_extensions.Literal
+        is_litteral = getattr(obj, '__values__', None) is not None
+
+        # typing_extensions.Final
+        is_final = getattr(obj, '__type__', None) is not None
+
+        # typing.Union/Tuple for old Python 3.5
+        is_union = getattr(obj, '__union_params__', None) is not None
+        is_tuple = getattr(obj, '__tuple_params__', None)
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:446]: Don't you miss the following?

```suggestion
        is_tuple = getattr(obj, '__tuple_params__', None) is not None
```

File: cloudpickle/cloudpickle.py (hunk 2 above)
Comment (pierreglaser) [reply cloudpickle/cloudpickle.py:989]: From a quick skim I have not seen anything about minor versions... :(

File: cloudpickle/cloudpickle.py (hunk 3 above)
Comment (pierreglaser) [reply cloudpickle/cloudpickle.py:446]: thanks.

**Comment by pierreglaser** [comment]
rebased.

File: cloudpickle/cloudpickle.py (hunk 2 above)
Comment (ogrisel) [reply cloudpickle/cloudpickle.py:989]: but have you tried to update the yaml file directly to point to 3.5.2 instead of 3.5? It's worth a blind try.

Comment (pierreglaser) [reply cloudpickle/cloudpickle.py:989]: let's try.

Comment (pierreglaser) [reply cloudpickle/cloudpickle.py:989]: Seems like it's not working. Maybe https://github.com/goanpeca/setup-miniconda?

**Comment by ogrisel** [comment]
Maybe it's not worth fighting with the CI too much to test Python 3.5.0... I will test this PR one more time locally.

**Comment by ogrisel** [comment]
Maybe it's not worth fighting with the CI too much to test Python 3.5.0... I will test this PR one more time locally.

**Comment by pierreglaser** [comment]
Cool.

**Comment by ogrisel** [comment]
How did you install Python 3.5.0 to run your tests? From source? I cannot find it with conda.

**Comment by pierreglaser** [comment]
let me know when if it works locally - if so, I'll clean up the CI additions in this PR.

**Comment by pierreglaser** [comment]
> How did you install Python 3.5.0 to run your tests? From source? I cannot find it with conda.

Using `miniconda2`.

**Comment by ogrisel** [comment]
Ok I managed to run the tests of this PR successfully using python:3.5.0 from docker.

**Comment by ogrisel** [comment]
From your last CI run:

```
ERROR: Package 'cloudpickle' requires a different Python: 2.7.17 not in '>=3.5'
```

That's an interesting error message ;) I am not sure what it means.

Anyways, feel free to rollback the CI changes and merge this PR to make the release.

**Comment by pierreglaser** [comment]
Great. I'm quickly testing 3.5.1 and 3.5.2 while I'm at it, and i'll merge.

**Comment by ogrisel** [comment]
Actually I made a mistake when testing with Python 3.5.0. I had forgotten to install the typing_extensions package. If I do so I get the following failure:

```python-traceback
___________________________________________________________________ CloudPickleTest.test_generic_extensions ____________________________________________________________________

self = <tests.cloudpickle_test.CloudPickleTest testMethod=test_generic_extensions>

    def test_generic_extensions(self):
        typing_extensions = pytest.importorskip('typing_extensions')
    
        objs = [
            typing_extensions.Literal,
            typing_extensions.Final,
            typing_extensions.Literal['a'],
            typing_extensions.Final[int],
        ]
    
        for obj in objs:
            depickled_obj = pickle_depickle(obj, protocol=self.protocol)
>           assert depickled_obj == obj

tests/cloudpickle_test.py:2137: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.5/site-packages/typing_extensions.py:653: in __eq__
    if not isinstance(other, Literal):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = typing_extensions.Literal, obj = typing_extensions.Literal

    def __instancecheck__(self, obj):
>       raise TypeError("Literal cannot be used with isinstance().")
E       TypeError: Literal cannot be used with isinstance().

/usr/local/lib/python3.5/site-packages/typing_extensions.py:624: TypeError

```

**Comment by pierreglaser** [comment]
What kind of can of worms is this.....

**Comment by pierreglaser** [comment]
The test suite passes locally on `cloudpickle 3.5.[0-6]` for me. `3.5.9` (the one used by the CI) passes too. `miniconda` does not seem to have `3.5.[7-8]` so it's harder for me to test those two versions.

Let's sleep on it until tomorrow morning and then release.

**Comment by ogrisel** [comment]
I confirm the tests now passe on my local 3.5.0. I will test each other 3.5.x version with docker and then review the diff.

**Comment by ogrisel** [comment]
For your information, here is the command line I use to run the tests with docker:

```bash
docker run --rm -v `pwd`:/io -ti python:3.5.2 \
    bash -c "cd /io/ && pip install -e . psutil pytest typing_extensions && pytest -v"
```

File: cloudpickle/cloudpickle.py [hunk 4]
```diff
@@ -432,10 +437,24 @@ def _extract_class_dict(cls):
 if sys.version_info[:2] < (3, 7):  # pragma: no branch
     def _is_parametrized_type_hint(obj):
         # This is very cheap but might generate false positives.
-        origin = getattr(obj, '__origin__', None)  # typing Constructs
-        values = getattr(obj, '__values__', None)  # typing_extensions.Literal
-        type_ = getattr(obj, '__type__', None)     # typing_extensions.Final
-        return origin is not None or values is not None or type_ is not None
+        # genral typing Constructs
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:440]: This fix was lost in the fight apparently:

```suggestion
        # general typing Constructs
```

File: cloudpickle/cloudpickle.py [hunk 5]
```diff
@@ -1295,20 +1340,17 @@ def _is_dynamic(module):
...
 
-def _make_typevar(name, bound, constraints, covariant, contravariant,
-                  class_tracker_id):
-    tv = typing.TypeVar(
+def _make_typevar(name, bound, constraints, covariant, contravariant):
+    return typing.TypeVar(
         name, *constraints, bound=bound,
         covariant=covariant, contravariant=contravariant
     )
-    return _lookup_class_or_track(class_tracker_id, tv)
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:1304]: Why don't you track `TypeVar` defintions anymore? This seems unrelated to the change to support early 3.5.x. 

Doesn't this change break ant test?

Edit: I see you removed `test_pickle_dynamic_typevar_tracking`.

Comment (ogrisel) [reply cloudpickle/cloudpickle.py:1304]: Also won't this break unpickling objects pickled with cloudpickle 1.4.0?

Comment (pierreglaser) [reply cloudpickle/cloudpickle.py]: `TyperVar` instances are not weakreferable in `Python 3.5.3`..

**Comment by ogrisel** [comment]
Ok I re-enabled dynamic `TypeVar` tracking and the tests still pass for all the 3 oldest versions of `3.5.x`. Will try others in the 3.5.3-3.5.9 range at random.

File: cloudpickle/cloudpickle.py (hunk 5 above)
Comment (pierreglaser) [reply cloudpickle/cloudpickle.py]: > Also won't this break unpickling objects pickled with cloudpickle 1.4.0?

We can always restore the `class_tracker_id` for backward compat.

File: tests/cloudpickle_test.py [hunk 6]
```diff
@@ -2081,20 +2081,32 @@ class C(typing.Generic[T]):
...
-                assert generic.__args__[0] is type_value
 
-                assert len(origin.__orig_bases__) == 1
-                ob = origin.__orig_bases__[0]
-                assert ob.__origin__ is typing.Generic
+                if sys.version_info >= (3, 5, 3):
+                    assert len(origin.__orig_bases__) == 1
+                    ob = origin.__orig_bases__[0]
+                    assert ob.__origin__ is typing.Generic
+                else:  # Python 3.5.[0-1-2]
```
Comment (ogrisel) [inline tests/cloudpickle_test.py:2091]:
```suggestion
                else:  # Python 3.5.[0-1-2], pragma: no cover
```

File: tests/cloudpickle_test.py [hunk 7]
```diff
@@ -2081,20 +2081,32 @@ class C(typing.Generic[T]):
...
+                else:
+                    assert len(generic.__parameters__) == 1
+                    assert generic.__parameters__[0] is type_value
                 assert len(ob.__parameters__) == 1
 
                 return "ok"
 
-            assert check_generic(C[int], C, int) == "ok"
-            assert worker.run(check_generic, C[int], C, int) == "ok"
+            # backward-compat for old Python3.5 versions that sometimes relies
```
Comment (ogrisel) [inline tests/cloudpickle_test.py:2105]:
```suggestion
            # backward-compat for old Python 3.5 versions that sometimes relies
```

**Comment by pierreglaser** [comment]
> Ok I re-enabled dynamic TypeVar tracking and the tests still pass for all the 3 oldest versions of 3.5.x

I'm not sure we can. See https://github.com/cloudpipe/cloudpickle/pull/361#discussion_r417117817

File: tests/cloudpickle_test.py [hunk 8]
```diff
@@ -2116,19 +2128,38 @@ def check_annotations(obj, expected_type):
...
+        def check_literal_equal(obj1, obj2):
+            assert obj1.__values__ == obj2.__values__
+            assert type(obj1) == type(obj2) == typing_extensions._LiteralMeta
+        literal_objs = [
+            typing_extensions.Literal, typing_extensions.Literal['a']
         ]
+        for obj in literal_objs:
+            depickled_obj = pickle_depickle(obj, protocol=self.protocol)
+            try:
+                # __eq__ does not work for Literal objects in early Python 3.5
```
Comment (ogrisel) [inline tests/cloudpickle_test.py:2143]: I think this test should inspect the Python version to know which assertions to check instead of using a try / except block.

File: tests/cloudpickle_test.py [hunk 9]
```diff
@@ -2116,19 +2128,38 @@ def check_annotations(obj, expected_type):
...
 
-        for obj in objs:
+        for obj in final_objs:
             depickled_obj = pickle_depickle(obj, protocol=self.protocol)
-            assert depickled_obj == obj
+            try:
+                # __eq__ does not work for Final objects in early Python 3.5
+                assert depickled_obj == obj
+            except Exception:
+                check_final_equal(obj, depickled_obj)
```
Comment (ogrisel) [inline tests/cloudpickle_test.py:2162]: Same remark here.

File: tests/cloudpickle_test.py (hunk 8 above)
Comment (pierreglaser) [reply tests/cloudpickle_test.py:2143]: Sure.

File: cloudpickle/cloudpickle.py (hunk 5 above)
Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: > TyperVar instances are not weakreferable in Python 3.5.3..

I re-ran the tests with old Python 3.5.x and they pass...

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Hum you are right they fail just for 3.5.3 ...

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: ok let me do a workaround for 3.5.3.
